staging/src/k8s.io/client-go/discovery/discovery_client.go

```diff
@@ -156,36 +174,139 @@ func apiVersionsToAPIGroup(apiVersions *metav1.APIVersions) (apiGroup metav1.API
+		}
+		apiGroupList.Groups = []metav1.APIGroup{apiGroup}
+	case acceptV2Beta1:
+		var aggregatedDiscovery apidiscovery.APIGroupDiscoveryList
+		err = json.Unmarshal(body, &aggregatedDiscovery)
+		if err != nil {
+			return nil, nil, err
+		}
+		apiGroupList, resourcesByGV = SplitGroupsAndResources(aggregatedDiscovery)
+	default:
+		return nil, nil, fmt.Errorf("Unknown discovery response content-type: %s", responseContentType)
 	}
 
-	// Get the groupVersions exposed at /apis
-	apiGroupList = &metav1.APIGroupList{}
-	err = d.restClient.Get().AbsPath("/apis").Do(context.TODO()).Into(apiGroupList)
+	return apiGroupList, resourcesByGV, nil
+}
+
+// downloadAPIs returns the discovery groups and (if aggregated format) the
+// discovery resources. The returned groups will always exist, but the
+// resources map may be nil.
+func (d *DiscoveryClient) downloadAPIs() (*metav1.APIGroupList, map[schema.GroupVersion]*metav1.APIResourceList, error) {
+	var responseContentType string
+	body, err := d.restClient.Get().
+		AbsPath("/apis").
+		SetHeader("Accept", acceptDiscoveryFormats).
+		Do(context.TODO()).
+		ContentType(&responseContentType).
+		Raw()
+	// Special error handling for 403 or 404 to be compatible with older v1.0 servers.
+	// Return empty group list to be merged with /api.
 	if err != nil && !errors.IsNotFound(err) && !errors.IsForbidden(err) {
-		return nil, err
+		return nil, nil, err
 	}
-	// to be compatible with a v1.0 server, if it's a 403 or 404, ignore and return whatever we got from /api
 	if err != nil && (errors.IsNotFound(err) || errors.IsForbidden(err)) {
-		apiGroupList = &metav1.APIGroupList{}
+		return &metav1.APIGroupList{}, nil, nil
 	}
 
-	// prepend the group retrieved from /api to the list if not empty
-	if len(v.Versions) != 0 {
-		apiGroupList.Groups = append([]metav1.APIGroup{apiGroup}, apiGroupList.Groups...)
+	apiGroupList := &metav1.APIGroupList{}
+	var resourcesByGV map[schema.GroupVersion]*metav1.APIResourceList
+	// Switch on content-type server responded with: aggregated or unaggregated.
+	switch responseContentType {
+	case acceptV1:
+		err = json.Unmarshal(body, apiGroupList)
+		if err != nil {
+			return nil, nil, err
+		}
+	case acceptV2Beta1:
+		var aggregatedDiscovery apidiscovery.APIGroupDiscoveryList
+		err = json.Unmarshal(body, &aggregatedDiscovery)
+		if err != nil {
+			return nil, nil, err
+		}
+		apiGroupList, resourcesByGV = SplitGroupsAndResources(aggregatedDiscovery)
+	default:
+		return nil, nil, fmt.Errorf("Unknown discovery response content-type: %s", responseContentType)
+	}
+
+	return apiGroupList, resourcesByGV, nil
+}
+
+// ServerGroups returns the supported groups, with information like supported versions and the
+// preferred version.
+func (d *DiscoveryClient) ServerGroups() (apiGroupList *metav1.APIGroupList, err error) {
+	groups, _, err := d.GroupsAndMaybeResources()
+	if err != nil {
+		return nil, err
 	}
-	return apiGroupList, nil
+	return groups, nil
```

Comment:
the signature of this function is still naming the output param `apiGroupList`. feels weird to just ignore that

Reply:
Fixed.